src/common/iso_8601.cc

```diff
@@ -0,0 +1,209 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <cctype>
+#include <chrono>
+#include <ctime>
+#include <cstdint>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <utility>
+
+#include "iso_8601.h"
+
+#include <iostream>
+
+namespace ceph {
+using std::chrono::duration_cast;
+using std::chrono::nanoseconds;
+using std::chrono::seconds;
+using std::setfill;
+using std::setw;
+using std::size_t;
+using std::stringstream;
+using std::string;
+using std::uint16_t;
+using std::uint64_t;
+
+using boost::none;
+using boost::optional;
+using boost::string_ref;
+
+using ceph::real_clock;
+using ceph::real_time;
+
+using sriter = string_ref::const_iterator;
+
+namespace {
+// This assumes a contiguous block of numbers in the correct order.
+uint16_t digit(char c) {
+  if (!(c >= '0' && c <= '9')) {
+    throw std::invalid_argument("Not a digit.");
+  }
+  return static_cast<uint64_t>(c - '0');
+}
+
+optional<real_time> calculate(tm& t, uint32_t n = 0) {
+  ceph_assert(n < 1000000000);
+  time_t tt = timegm(&t);
+  if (tt == static_cast<time_t>(-1)) {
+    return none;
+  }
+
+  return real_clock::from_time_t(tt) + nanoseconds(n);
+}
+}
+
+optional<real_time> from_iso_8601(const string_ref s,
+				  const bool ws_terminates) noexcept {
```

Comment:
Yeah, there's a modeline in most of the Ceph source files that says to replace runs of spaces with tabs.

*Though* my personal preference would be to just go all-spaces. The Linux kernel pattern of having one tab=one indentation level at least makes some sort of sense. Having two spaces be one indentation level and also we replace runs of spaces with tabs for some reason seems kind of...dumb...